dev/ProgressBar/ProgressBar.cpp

```diff
@@ -0,0 +1,150 @@
+
+    SetDefaultStyleKey(this);
+
+    // NOTE: This is necessary only because Value isn't one of OUR properties, it's implemented in RangeBase.
+    // If it was one of ProgressBar's properties, defined in the IDL, you'd do it differently (see IsIndeterminate).
+    RegisterPropertyChangedCallback(winrt::RangeBase::ValueProperty(), { this, &ProgressBar::OnRangeBasePropertyChanged });
+    RegisterPropertyChangedCallback(winrt::RangeBase::MinimumProperty(), { this, &ProgressBar::OnRangeBasePropertyChanged });
+    RegisterPropertyChangedCallback(winrt::RangeBase::MaximumProperty(), { this, &ProgressBar::OnRangeBasePropertyChanged });
+
+    SetValue(s_TemplateSettingsProperty, winrt::make<::ProgressBarTemplateSettings>());
+}
+
+void ProgressBar::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected{ *this };
+
+    // NOTE: Example of how named parts are loaded from the template. Important to remember that it's possible for
+    // any of them not to be found, since devs can replace the template with their own.
+
+    m_layoutRoot.set(GetTemplateChildT<winrt::Grid>(s_LayoutRootName, controlProtected));
+    m_progressBarIndicator.set(GetTemplateChildT<winrt::Rectangle>(s_ProgressBarIndicatorName, controlProtected));
+
+    UpdateStates();
+}
+
+void ProgressBar::OnRangeBasePropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    // NOTE: This hits when the Value property changes, because we called RegisterPropertyChangedCallback.
+
+    SetProgressBarIndicatorWidth();
+}
+
+void ProgressBar::OnIsIndeterminatePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    // NOTE: This hits when IsIndeterminate changes because we set MUX_PROPERTY_CHANGED_CALLBACK to true in the idl.
+
+    UpdateStates();
+}
+
+void ProgressBar::OnShowPausedPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateStates();
+}
+
+void ProgressBar::OnShowErrorPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateStates();
+}
+
+void ProgressBar::UpdateStates()
+{
+    m_sizeChangedRevoker.revoke();
+
+    if (ShowError())
+    {
+        winrt::VisualStateManager::GoToState(*this, s_ErrorStateName, true);
+    }
+    else if (ShowPaused() && IsIndeterminate())
+    {
+        winrt::VisualStateManager::GoToState(*this, s_ErrorStateName, true); // Paused-Indeterminate state same visual treatment as Error state
+    }
+    else if (ShowPaused())
+    {
+        winrt::VisualStateManager::GoToState(*this, s_PausedStateName, true);
+    }
+    else if (IsIndeterminate())
+    {
+        UpdateWidthBasedTemplateSettings(nullptr, nullptr);
+        winrt::VisualStateManager::GoToState(*this, s_IndeterminateStateName, true);
+        m_sizeChangedRevoker = this->SizeChanged(winrt::auto_revoke, { this, &ProgressBar::UpdateWidthBasedTemplateSettings });
+    }
+    else if (!IsIndeterminate())
+    {
+        SetProgressBarIndicatorWidth();
```

Comment:
>SetProgressBarIndicatorWidth [](start = 8, length = 28)

Does this fix the bug where leaving the indeterminate mode we saw last week?

Reply:
Yes, it resets the width of ProgressBar to the value in Determinate state